airbyte-cdk/bulk/core/load/src/main/kotlin/io/airbyte/cdk/task/DestinationTaskLauncher.kt

```diff
@@ -188,47 +188,109 @@ class DefaultDestinationTaskLauncher(
     }
 }
 
+/**
+ * The exception handler takes over the workflow in the event of an exception. Its contract is
+ * * provide a wrapper that directs exceptions to the correct handler (by task type)
+ * * close the task runner when the cleanup workflow is complete
+ *
+ * Handling works as follows:
+ * * a failure in a sync-level task (setup/teardown) triggers a fail sync task
+ * * a failure in a stream task triggers a fails stream task THEN a fail sync task
+ * * the wrappers will skip tasks if the sync/stream has already failed
+ */
 @Singleton
 @Secondary
-class DefaultDestinationTaskLauncherExceptionHandler : DestinationTaskLauncherExceptionHandler {
+class DefaultDestinationTaskLauncherExceptionHandler(
+    private val taskRunner: TaskRunner,
+    private val syncManager: SyncManager,
+    private val failStreamTaskFactory: FailStreamTaskFactory,
+    private val failSyncTaskFactory: FailSyncTaskFactory,
+) : DestinationTaskLauncherExceptionHandler {
     class SyncTaskWrapper(
         private val exceptionHandler: DestinationTaskLauncherExceptionHandler,
+        private val syncManager: SyncManager,
         private val innerTask: SyncTask,
     ) : Task {
+        val log = KotlinLogging.logger {}
+
         override suspend fun execute() {
+            if (!syncManager.isActive()) {
+                val result = syncManager.awaitSyncResult()
+                if (result is SyncSuccess) {
+                    throw IllegalStateException(
+                        "Task $innerTask run after sync has succeeded. This should not happen."
+                    )
+                }
+                log.info { "Sync terminated, skipping task $innerTask." }
+            }
+
             try {
                 innerTask.execute()
             } catch (e: Exception) {
                 exceptionHandler.handleSyncFailure(e)
             }
         }
+
+        override fun toString(): String {
+            return "SyncTaskWrapper(innerTask=$innerTask)"
+        }
     }
 
     class StreamTaskWrapper(
         private val exceptionHandler: DestinationTaskLauncherExceptionHandler,
+        private val syncManager: SyncManager,
         private val innerTask: StreamTask,
     ) : SyncTask {
+        val log = KotlinLogging.logger {}
+
         override suspend fun execute() {
+            // Stop dispatching tasks if the stream has been killed by a failure elsewhere.
+            // Specifically fail if the stream was marked succeeded: we should not be in this state.
```

Comment:
reminding myself: this state is illegal because we don't mark the stream as succeeded until all the batches end up in a COMPLETED state, and therefore there shouldn't be any more batches to process (and therefore no more tasks?)

Reply:
Right. Specifically, the last stream task (CloseStreamTask) marks the stream complete. So we should not be launching any more stream tasks for that stream after it runs.

High-level, the contract is "don't launch stream tasks after the stream is complete."

This is also why `StreamLoader::close` and `DestinationWriter::teardown` take nullable failure types and not `Result` types: there won't be a result until after they complete.